server/server.go

```diff
@@ -29,47 +32,81 @@ type ServerConfig struct {
 
 // Analyzer is a struct of analyzer client and config
 type Analyzer struct {
-	Client AnalyzerClient
+	Client lookout.AnalyzerClient
 	Config AnalyzerConfig
 }
 
-type reqSent func(client AnalyzerClient, settings map[string]interface{}) ([]*Comment, error)
+type reqSent func(client lookout.AnalyzerClient, settings map[string]interface{}) ([]*lookout.Comment, error)
 
 // Server implements glue between providers / data-server / analyzers
 type Server struct {
-	watcher    Watcher
-	poster     Poster
-	fileGetter FileGetter
+	watcher    lookout.Watcher
+	poster     lookout.Poster
+	fileGetter lookout.FileGetter
 	analyzers  map[string]Analyzer
+	eventOp    store.EventOperator
+	commentOp  store.CommentOperator
 }
 
 // NewServer creates new Server
-func NewServer(w Watcher, p Poster, fileGetter FileGetter, analyzers map[string]Analyzer) *Server {
-	return &Server{w, p, fileGetter, analyzers}
+func NewServer(w lookout.Watcher, p lookout.Poster, fileGetter lookout.FileGetter,
+	analyzers map[string]Analyzer, eventOp store.EventOperator, commentOp store.CommentOperator) *Server {
+	return &Server{w, p, fileGetter, analyzers, eventOp, commentOp}
 }
 
 // Run starts server
 func (s *Server) Run(ctx context.Context) error {
 	// FIXME(max): we most probably want to change interface of EventHandler instead of it
-	return s.watcher.Watch(ctx, func(e Event) error {
+	return s.watcher.Watch(ctx, func(e lookout.Event) error {
 		return s.handleEvent(ctx, e)
 	})
 }
 
-func (s *Server) handleEvent(ctx context.Context, e Event) error {
+func (s *Server) handleEvent(ctx context.Context, e lookout.Event) (err error) {
+	logger := log.With(log.Fields{
+		"event-type": e.Type(),
+		"event-id":   e.ID(),
+	})
+
+	status, err := s.eventOp.Save(ctx, e)
+	if err != nil {
+		logger.Errorf(err, "can't save event to database")
+		return
+	}
+
+	if status == models.EventStatusProcessed {
+		logger.Infof("event successfully processed, skipping...")
+		return
+	}
+
+	defer func() {
```

Comment:
If we have only one return statement, do we need a defer?
Wouldn't it be more clear if this code was after the switch?

Reply:
true. Thank you. I believe before I was returning inside switch that's why I needed defer. But not it will be much better to move after switch.